EIPS/eip-131.md

```diff
@@ -0,0 +1,83 @@
+---
+eip: 131
+title: Add precompiled BLAKE2b contract
+author: Tjaden Hess (@tjade273), Jay Graber, (@carterpy), James Hancock (@madeoftin)
+discussions-to: https://ethereum-magicians.org/t/blake2b-f-precompile/3157
+status: Draft
+type: Standards Track
+category: Core
+created: 2016-06-30
+---
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+Special Projects wishes greater Ethereum interopability with Zcash, IPFS, and Handshake. Coincidentally, all of these projects use the Blake2 hash function. So, this is a request to add the Blake2b F precompile 11 to the Istanbul hardfork. The first special project here will probably be creating a wrapped ZEC (WZEC) within Ethereum as well as wrapped Ether within Zcash. After that, some yet-to-be-determined bridge architecture will allow Ethereum to benefit from Zcash’s shielded transactions.
+
+## Abstract
+
+This EIP introduces a new precompiled contract which implements the BLAKE2b cryptographic hashing algorithm.
+
+## Motivation
+
+BLAKE2b is a useful addition to Ethereum's cryptographic primitives. BLAKE2b is faster and more efficient than the SHA hash functions, making it a good complement or alternative to SHA-3. The BLAKE2b algorithm is highly optimized for 64-bit CPUs, and is faster than MD5 on modern processors. Its inclusion will also improve interoperability between the Zcash blockchain and the EVM, and assist in future integration with IPFS, as they are working on switching their default hash function to BLAKE2b.
+
+One BLAKE2 digest in Solidity, (see https://github.com/tjade273/eth-blake2/blob/optimise_everything/contracts/blake2.sol) currently requires `~480,000 + ~90*INSIZE` gas, and a single verification of an [Equihash](https://www.cryptolux.org/images/b/b9/Equihash.pdf) PoW verification requires 2<sup>5</sup> to 2<sup>7</sup> iterations of the hash function, making use cases such as verification of Zcash block headers prohibitively expensive. However, an efficient implementation of BLAKE2b could allow verification of the Equihash PoW used in Zcash, making a BTC Relay - style SPV client possible on Ethereum.
```

Comment:
So what is currently required for Equihash verification? `480000 + 90 * (2**7 * 128)` which is ~2M gas?

Reply:
I believe you're misreading. The verification cost for Equihash is `2**k` invocations of the hash function (and `2**k` XORs)

Reply:
I'm not sure why are you writing `2**7`. The `128` looks much less scary.

Reply:
`k` is a parameter, so we write `2**k`. Zcash uses Equihash with `k = 9`

Reply:
You really starts annoying me. If ZCash uses `k = 9` WHY have you written its from 5 to 7?

Reply:
> WHY have you written its from 5 to 7?

He hasn't written that, AFAIK @prestwich didn't author the EIP

> You really starts annoying me

This isn't productive, let's stick to the work and leave feelings at the door

> I believe you're misreading.

@prestwich considering your comments, does the EIP as it stands adequately address gas pricing?

Reply:
It should be modified to specify that Zcash uses Equihash200,9 and will need `2**9` iterations of blake2b (and `2**9` XORs. It might be good to add some napkin math for that usecase

For a single header we expect costs to be dominated by calldata and blake evaluation
-  **calldata** ~1.3kB of calldata at 68/byte, which is about 88,400 gas (may be reduced by EIP2048)
- **blake (pre-EIP)** 512 evaluations of blake2b at 480,000/eval is about 245,760,000
- **blake (post-EIP)** 512 evaluations at ~36/eval (extra block for personalization/keying), or 18,432 gas

Reply:
Instead of napkin math, it would be good to just add an example solidity contract that implements the primary usecase by calling the precompile.

Reply:
> blake (pre-EIP) 512 evaluations of blake2b at 480,000/eval is about 245,760,000

A source/citation of the pre-EIP implementation, or reproducible benchmark for these numbers would be good too.

Reply:
tjade273 demo'd Blake2 in solidity [in an earlier EIP issue](url). His .sol implementation[ is here](https://github.com/tjade273/eth-blake2/blob/optimise_everything/contracts/blake2.sol).

Reply:
I want to see Dave and Matt do an implementation of go-ethereum with the precompile in it and a .sol contract that uses that precompile to show the delta but like, the same way I want a new season of Rick and Morty.

Reply:
where is the fan page?

Reply:
Since that original code was posted a lot has changed. For example opcodes were repriced and shift instructions were introduced. It would be nice to see the gas cost of blake2 today. @cdetrio has a modified version of the blake2 solidity implementation and mentioned the some new costs on an ACD call 4-5 weeks ago.

Reply:
Instead of shifts your linked blake2 implementation uses exponentiation which is (extremely) expensive.

Reply:
hey @axic can you link to @cdetrio's implementation? Does he have it in a public repo?

Reply:
> I want to see James and Matt do an implementation of go-ethereum with the precompile in it and a .sol contract that uses that precompile to show the delta but like, the same way I want a new season of Rick and Morty.

I don't think its an unreasonable ask. This was done for the ECMUL/ECPAIRING precompiles before Byzantium (including an example solidity contract that used the ECPAIRING precompile to verify a snark proof that came from a zcash tx).

> Can we say Zac's code is the best smart contract implementation with the current EVM we're going to see and use that for @cdetrio's benchmark? Zac got it to 6,821,545 gas with pre-Istanbul EVM.

Yes, Zac said there probably isn't much room left for optimizing the huff implementation (I believe him). What isn't fully clear (to me) is whether the 6.8 million gas number is sufficient to verify a block. And if it is, why is 245 million cited above instead. What's annoying is to see people going around in circles here and on ACD calls with napkin math. Would be good if some champion could just write the solidity code to verify a zcash block or whatever the usecase for the precompile is (again, like was done for the snark precompiles).

Reply:
> I don't think its an unreasonable ask.

We're working on it- our team has focused on the `F` precompile first considering where [that discussion went](https://github.com/ethereum/EIPs/issues/152). I'll have an EIP PR for `F` out today where we link to our initial implementation, then we'll circle back here

Reply:
> Would be good if some champion could just write the solidity code to verify a zcash block or whatever the usecase for the precompile is (again, like was done for the snark precompiles).

Agreed, there will be benchmarks- first take will be for gas estimation of `F`, then another for `blake2b`. Full Zcash header verification is a stretch though we'll see what we can do.